File: src/main/java/com/yauhenl/poe/service/BaseMongoService.java
package com.yauhenl.poe.service;

import com.mongodb.client.FindIterable;
import com.mongodb.client.MongoCollection;
import com.mongodb.client.MongoDatabase;
import com.mongodb.client.model.UpdateOptions;
import org.bson.Document;
import org.bson.types.ObjectId;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Consumer;

import static com.mongodb.client.model.Filters.eq;
import static com.mongodb.client.model.Filters.in;

public abstract class BaseMongoService {

    protected MongoCollection<Document> mongoCollection;

    public BaseMongoService(MongoDatabase mongoDatabase, String mongoCollectionName) {
        mongoCollection = mongoDatabase.getCollection(mongoCollectionName);
        if (mongoCollection != null && mongoCollection.count() == 0) {
            initCollection();
        }
    }

    public void initCollection() {
    }

    public void drop() {
        mongoCollection.drop();
    }

    public void insertOne(Document document) {
        mongoCollection.insertOne(document);
    }

    public void insertMany(List<Document> documents) {
        mongoCollection.insertMany(documents);
    }

    public void replaceOne(Document document) {
        mongoCollection.replaceOne(eq("_id", document.get("_id")), document, new UpdateOptions().upsert(true));
    }

    public void deleteOne(Document document) {
        mongoCollection.deleteOne(document);
    }

    public void truncate() {
        mongoCollection.deleteMany(new Document());
    }

    public Object getMongoId(Document doc) {
        return doc.get("_id");
    }

    public void deleteByIds(List<Object> ids) {
        mongoCollection.deleteMany(in("_id", ids));
    }

    public Document findById(Object id) {
        return mongoCollection.find(eq("_id", id)).first();
    }

    public Document findFirstByProperty(String key, Object value) {
        return mongoCollection.find(eq(key, value)).first();
    }

    public FindIterable<Document> find() {
        return mongoCollection.find();
    }

    public long count() {
        return mongoCollection.count();
    }

    public Document findFirst() {
        return mongoCollection.find().first();
    }

    public void updateOneFieldById(String id, String fieldName, Object value) {
        mongoCollection.updateOne(eq("_id", new ObjectId(id)), new Document("$set", new Document(fieldName, value)));
    }

    protected <T> List<T> consumeToList(FindIterable<T> input) {
        List<T> result = new ArrayList<>();
        input.forEach((Consumer<T>) result::add);
        return result;
    }
}

File: src/main/kotlin/com/yauhenl/poe/service/BaseMongoService.kt
package com.yauhenl.poe.service

import com.mongodb.client.MongoCollection
import com.mongodb.client.MongoDatabase
import com.mongodb.client.model.Filters.`in`
import com.mongodb.client.model.Filters.eq
import com.mongodb.client.model.UpdateOptions
import org.bson.Document
import com.sun.xml.internal.ws.model.RuntimeModeler.getNamespace
import java.util.stream.Collectors
import jdk.nashorn.internal.objects.NativeArray.forEach
import java.util.*
import java.util.function.Consumer






abstract class BaseMongoService(mongoDatabase: MongoDatabase, mongoCollectionName: String) {
    protected var mongoCollection: MongoCollection<Document>

    open fun initCollection() {
    }

    fun drop() = mongoCollection.drop()

    fun insertOne(document: Document) = mongoCollection.insertOne(document)

    fun insertMany(documents: List<Document>) = mongoCollection.insertMany(documents)

    fun replaceOne(document: Document) = mongoCollection.replaceOne(eq("_id", document["_id"]), document, UpdateOptions().upsert(true))

    fun deleteOne(document: Document) = mongoCollection.deleteOne(document)

    fun truncate() = mongoCollection.deleteMany(Document())

    fun deleteByIds(ids: List<Any>) = mongoCollection.deleteMany(`in`("_id", ids))

    fun getMongoId(doc: Document) = doc["_id"]

    fun findById(id: Any) = mongoCollection.find(eq("_id", id)).first()

    fun findFirstByProperty(key: String, value: Any) = mongoCollection.find(eq(key, value)).first()

    fun find() = mongoCollection.find()

    fun count() = mongoCollection.count()

    fun findFirst() = mongoCollection.find().first()

    init {
        mongoCollection = mongoDatabase.getCollection(mongoCollectionName)
        if (mongoCollection.count().equals(0)) {
            initCollection()
        }
    }
}